app/vmstorage/servers/vmselect.go

```diff
@@ -253,16 +253,16 @@ func checkTimeRange(s *storage.Storage, tr storage.TimeRange) error {
 	}
 }
 
-func getMaxMetrics(sq *storage.SearchQuery) int {
-	maxMetrics := sq.MaxMetrics
-	maxMetricsLimit := *maxUniqueTimeseries
-	if maxMetricsLimit <= 0 {
-		maxMetricsLimit = GetMaxUniqueTimeSeries()
+func getMaxMetrics(searchQueryLimit int) int {
+	if searchQueryLimit <= 0 {
+		// use vmstorage maxUniqueTimeseries if no limit is passed by the client.
+		return GetMaxUniqueTimeSeries()
 	}
-	if maxMetrics <= 0 || maxMetrics > maxMetricsLimit {
-		maxMetrics = maxMetricsLimit
+	// if `-search.maxUniqueTimeseries` is explicitly set in vmstorage, client cannot exceed the limit.
```

Comment:
```suggestion
	// searchQueryLimit  can exceed value explicitly set to `-search.maxUniqueTimeseries`
```

this func is unaware of client/server terms